Add a simple catch-all implementation of the metadata URI interface

KaiRo-at opened the pull request:
As mentioned in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2014#issuecomment-565499083 this is a simple implementation of the metadata URI part of ERC1155.

I'm still a bit confused on how to do tests for OpenZeppelin (this is my first PR to the project), so I have not added tests in this PR for now. Help would be appreciated.

frangio commented:
Thanks @KaiRo-at! You should add a new file in the `test/token/ERC1155` directory for your new contract. It should be similar to `ERC1155.test.js` but testing only your specific behavior.

If you have any specific doubts please ask!

frangio commented:
One more thing. We recently migrated our tests to [OpenZeppelin Test Environment](https://github.com/OpenZeppelin/openzeppelin-test-environment). The `feature-erc1155` branch had forked before that, so it was behind with these changes. I have just updated it. Please update your branch and write your tests using the new tooling!

Keep in mind Mocha's `.only` modifier to run only the tests that are relevant to this PR while you're writing them: `describe.only('ERC1155Metadata...`.

KaiRo-at commented:
I just added a test to my repo, but I issues running tests, I get an `Error: Cannot find module 'openzeppelin-test-helpers'` when running `npm test`, how can I fix that?
`npm run lint` runs fine but there are warnings in `ERC1155.sol` which are not my fault. ;-)

nventuro commented:
Hello @KaiRo-at! From your description, it sounds like your import statement is wrong: you should import `@openzeppelin/test-helpers`, not `openzeppelin-test-helpers`. If that still doesn't work, please share a snippet or push your commits here so we can take a look.

And yes, feel free to ignore linting errors for now :)

KaiRo-at commented:
> From your description, it sounds like your import statement is wrong: you should import `@openzeppelin/test-helpers`, not `openzeppelin-test-helpers`.

I guess so - I copied from the other ERC1155 tests on that feature branch and it looks like it's wrong there as well... ;-)
And thanks for reopening this one!

nventuro commented:
Looks like you're right! The feature branch was created before we migrated our testing setup to use [Test Environment](https://full-site--openzeppelin-docs.netlify.com/test-environment/0.1/).

I updated the branch to bring the ERC1155 tests in line with the old ones, and then updated your own branch to use those. You should now be able to get your tests working! 

Just remember to run `npm install` first, so that the new dependencies are installed.

KaiRo-at commented:
Thanks for that, needed a bit of merging as I was working in parallel, but I should have tests up and running now. :)
The only detail is that I don't know how to get a receipt for the contract .new() to test for the event being emitted from the contructor as well (it's using the same function as being called and tested later though) - otherwise I think this should cover the whole functionality of the added contract/interface.

nventuro commented:
>The only detail is that I don't know how to get a receipt for the contract .new() to test for the event being emitted from the contructor as well

Have you tried using [`expectEvent.inConstruction`](https://docs.openzeppelin.com/test-helpers/0.5/api#inconstruction)? You should be able to use it as follows:

```javascript
const contract = await MyContract.new();
await expectEvent.inConstruction(contract, 'Foo', { value: 'bar' });
```

KaiRo-at commented:
> Have you tried using [`expectEvent.inConstruction`](https://docs.openzeppelin.com/test-helpers/0.5/api#inconstruction)?

Thanks, that's great! From what I see, everything should be complete for adding the URI feature now. :)

nventuro commented:
Sorry @KaiRo-at for taking so long to review this! I'll take a look tomorrow.

KaiRo-at commented:
Erm, @nventuro - I don't think this should have been closed...

nventuro commented:
It shouldn't have, no - I'm reopening it. I'm very sorry @KaiRo-at, I've been working on the [v3.0 pending changes](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2086) and haven't managed to get around to reviewing 1155-related PRs yet - the same thing applies to https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2107.

I'll take a look at this as soon as I can.

KaiRo-at commented:
Thanks, I also worked on getting the ERC1155 branch ported to 0.6, see #2130

KaiRo-at commented:
I moved the 0.6 porting from #2130 over here, so now tests for this will fail and only work fine once rebased on a state after that has been merged to the feature branch. I also squashed a few intermediate commits so the commits are a bit more logical.

KaiRo-at commented:
This is waiting for #2130  and for review.

nventuro commented on contracts/token/ERC1155/ERC1155MetadataURICatchAll.sol:
```diff
@@ -0,0 +1,46 @@
+     */
+    function _setURI(string memory newuri) internal virtual {
+        _uri = newuri;
+        emit URI(_uri, 0);
```
This seems sensible, but I couldn't find anything in the specification about the event's `id` argument in cases such as this. Given that an `id` value of 0 is valid, perhaps we shouldn't emit the event at all? The spec seems to allow this:

>Changes to the URI MUST emit the URI event if the change can be expressed with an event (i.e. it isn’t dynamic/programmatic).
>
>    An implementation MAY emit the URI event during a mint operation but it is NOT mandatory. An observer MAY fetch the metadata uri at mint time from the uri function if it was not emitted.
>
>The uri function SHOULD be used to retrieve values if no event was emitted.

KaiRo-at replied:
Oh, you mean "it isn't dynamic" says that the `{id}` pattern doesn't need it? I interpreted it as it can only be omitted if e.g. the URI is built by a function that depends on, say, other fields and the actually returned URI changes dynamically - but in our case, it can be expressed by an event, and the actually returned URI doesn't change dynamically, it changes only with that call. I have wondered about the ID in that event as well though, unfortunately, the spec doesn't define this case. Maybe we should ask there.

nventuro replied:
No, I agree with your interpretation and believe this case might be missing from the spec. Emitting an event feels wrong though because there is no invalid values for the `id` field we can use to signal what is going on. Asking the spec authors sounds like the way to go.

nventuro replied:
At any rate, given the spec allows for no events to be emitted (like in the case of dynamic URIs), I think it makes sense to not emit anything here.

KaiRo-at replied:
> At any rate, given the spec allows for no events to be emitted (like in the case of dynamic URIs), I think it makes sense to not emit anything here.

We definitely interpret the spec differently there, as I read it as non-optional, with the only exception being where it is not static and not possible to emit the event - and it is static in our case and very much possible to emit it. But if you as the reviewer(s) prefer that, I'll remove it even if I disagree. ;-)

nventuro replied:
We've also discussed this bit with @frangio and agree that we should not emit any events. The spec clearly allows for situations where no events have been emitted, as seen in these statements:

>The uri function SHOULD be used to retrieve values if no event was emitted.
>The uri function MUST return the same value as the latest event for an _id if it was emitted.

nventuro commented on contracts/token/ERC1155/ERC1155MetadataURICatchAll.sol:
```diff
@@ -0,0 +1,46 @@
+     * @param newuri New URI to be set
+     */
+    function _setURI(string memory newuri) internal virtual {
+        _uri = newuri;
```
What do you think about a runetime check validating the existance of the `{id}` substring? @frangio this is an interesting use case for string utils.

KaiRo-at replied:
Well, you could in theory set a completely static URI that points to the same JSON document for all IDs. There is nothing in the metadata that specifically has to be different for different IDs or unique to one ID. In practice, users will probably always want the ID in the URI, but it's not completely required.
If there is a string util for detecting if a string is contained in another, I'd be fine with adding such a check but I don't want to code that test myself. ;-)

nventuro replied:
True, though I'm not sure a constant URI for all ids is very useful :stuck_out_tongue: I was just wondering if people would consider this helpful or annoying.

frangio replied:
It's clearly optional in the EIP so I wouldn't enforce it.
>If the string {id} exists in any URI, clients MUST replace this with the actual token ID in hexadecimal form.

KaiRo-at commented:
> Sorry it took so long to get here @KaiRo-at, but the time for ERC1155 has finally come!
> 
> Given the work we've done on including optional extensions into the base implementation in ERC20 and ERC721, and the fact that URIs are an almost mandatory feature of NFTs, I was wondering if we could figure out a way to include this in the default ERC1155 contract, perhaps following a scheme similar to the [baseURI](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/1970) from ERC721. What if we had a 'fallback' URI, allowing users to replace it with a token-id specific one if they desire? Concatenating the token id doesn't seem necessary, given the string substitution requirements of the spec.

We could do that. That said, given that the ERC1155 spec contains this `{id}` thing directly in the standard, I'd guess it's preferred to use that there (though the URIs don't look as nice for sure and the web service behind it needs to do a bit more work to parse it into a normal number). Also, the question is how we'd differentiate between the `baseURI` scheme and the scheme with `{id}`.

nventuro commented:
>Also, the question is how we'd differentiate between the baseURI scheme and the scheme with {id}.

Sorry, I should've been clearer. I wasn't suggesting using `baseURI` directly, but rather the idea from that PR where there is a 'default' URI that can be overridden by setting a specific one for the token. Given string substitution in 1155, there seems to be no need for the baseURI concatenation we have in 721, but we might want to implement this as follows:

```
function _setDefaultURI();
function _setURI(id);

function uri(id) view {
  if uris[id]
    return uris[id];
  else
    return defaultURI;
}
```

KaiRo-at commented:
Sorry I have been silent on this and the other issue for a bit, but we just shipped (and sold out) our Crypto stamp 2 presale today using the code from here and the other ERC1155 contributions: https://etherscan.io/address/presale.cs2.cryptostamp.eth#code
I hope I'll get back to looking into this in the next days now that the immediate release stress is gone.

nventuro commented:
No worries, and congratulations on your release! :tada:

KaiRo-at commented:
> ```
> function uri(id) view {
>   if uris[id]
>     return uris[id];
>   else
>     return defaultURI;
> }
> ```

I definitely can do that. I concentrated on the single catch-all solution for now due to https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2014#issuecomment-566202908 and I was planning to look into extending that for per-token URIs once ERC1155 was on master if my time allowed for it. But I can look into the right here as well.

nventuro commented:
Good point, thanks for reminding me of that. I've discussed this with @frangio and came to the conclusion that, given there's a clear pattern on how to use URIs with string substituion, we can do without `_setDefaultURI()` and can just have a generic `_setUri()`.

nventuro commented:
@KaiRo-at we're aiming to release v3.1 soon and to include initial support for ERC1155 there, so if you don't mind I'll go ahead and make the final changes required for this PR myself :)

nventuro commented:
I added the optional extension to `ERC1155`, inline with how we've implemented other token standards. This let me simplify the tests and mocks quite a bit.

I additionally changed some `abstract contract`s into `interface`s, using the new interface inheritance mechanism, which resulted in some superficial changes in some files.

KaiRo-at commented:
> @KaiRo-at we're aiming to release v3.1 soon and to include initial support for ERC1155 there, so if you don't mind I'll go ahead and make the final changes required for this PR myself :)

Sure, I don't feel territorial about this, happy I could push it along some way. Sorry I'm pretty short on time recently (due to a lot of ongoing work for Crypto stamp 2), I'm happy this work here gets picked up and driven into a release!

nventuro commented:
Reviewed offline by @frangio.
